Update `only_if` to allow user specified messages.

miah opened the pull request:
Updates only_if to allow a user to specify a message to say why it was skipped. If the user does not supply a message the default
message is used.

```
control 'nutcracker-connect-redis-001' do
  impact 1.0
  title 'Check if nutcracker can pass commands to redis'
  desc 'execute redis-cli set key command, to check connectivity of the service'

  only_if('redis is not installed.') do
    command('redis-cli').exist?
  end

  describe command('redis-cli SET test_inspec "HELLO"') do
    its(:stdout) { should match(/OK/) }
  end
end
```

```
  ↺  nutcracker-connect-redis-001: Check if nutcracker can pass commands to redis
     ↺  Skipped control due to only_if condition: redis is not installed.
```

Ref: https://github.com/inspec/inspec/issues/3158
Signed-off-by: Miah Johnson <miah@chia-pet.org>

Fixes: https://github.com/inspec/inspec/issues/1900

miah committed:
This commit updates only_if to allow a user to specify a message to say
why it was skipped. If the user does not supply a message the default
message is used.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

jquick reviewed (commented):
Really nice enhancement on this. I think we will also need to update: https://github.com/inspec/inspec/blob/ca1746debb0a3a1c72394e63b991ad8c111eb478/lib/inspec/rule.rb#L177 to set the `skip_rule[:result]` and maybe add a test for this case.

miah committed:
Correct rubocop issue
Set {:result} on skip_rule instance variable.
check for msg[:result]

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Allow passing a message to set_skip_rule. Previously, the `value` passed
to set_skip_rule could be a boolean, or a message. Now value should
always be a boolean, and if a message is needed one can be passed and
will be set.
Allow only_if to take a message during control_eval DSL.
Add test for only_if(message).

Signed-off-by: Miah Johnson <miah@chia-pet.org>

jerryaldrichiii commented on lib/inspec/rule.rb:
```diff
@@ -183,8 +185,12 @@ def self.merge_count(rule)
-      return checks(rule) unless msg
-      msg = 'Skipped control due to only_if condition.' if msg == true
+      return checks(rule) unless msg[:result].eql?(true)
+      if msg[:result].eql?(true) && msg[:message].nil?
```
You can remove the first criterion since we cannot reach here if `msg[:result]` is false. I'd also modify it a bit by changing `msg` to `skip_check` (or some better name...) then do the following:

```ruby
if skip_check[:message]
  msg = "Skipped control due to only_if condition: #{msg[:message]}"
else
  msg = "Skipped control due to only_if condition."
end

jerryaldrichiii reviewed (changes requested):
Looks great @miah only one small suggestion.

clintoncwolfe commented on test/unit/profiles/profile_context_test.rb:
```diff
@@ -117,6 +117,7 @@ def load(call)
     describe 'global only_if' do
       let(:if_true) { "only_if { true }\n" }
```
For my part, having all of these brief `let`s containing short Strings is a bit confusing; if the lets were inlined, I think it would read clearer.  But that was already there.

clintoncwolfe reviewed (approved):
Code looks good; tests are good on coverage but read a bit unclear too me.  Not a blocker. Thanks, @miah!

miah committed:
Remove conditional checks for true and reverse if conditional on
Inline if_false_message into test.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Create a new variable 'msg' to store the message string returned when
skipped.

Signed-off-by: Miah Johnson <miah@chia-pet.org>
